README.md
# Publisher convert

[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)

Кароч, это конвертер картинок для Publisher'а.
Очень часто возникает ситуация, что картинки не нравятся клиенту.
Этот запил призван облегчить процесс конвертирования.

## Установка

Используй composer и все.

``` bash
$ php composer.phar require --prefer-dist pers1307/convert "dev-master"
```

## Инструкция

Пока что в пакете можно найти только сборник рецептов.

Добавление новой картинки в коллекцию, или замена старой

Быстрое решение

``` php
require_once 'console.php';

$query = new MSTable('{catalog_items}');
$query->setFields(['*']);
$items = $query->getItems();

$conf = array(115, 115, true);

foreach ($items as $key => &$item) {
    $buf = unserialize($item['gallery']);

    foreach ($buf as $key2 => &$elem) {

        if (file_exists(DOC_ROOT . $elem['path']['original'])) {
            $result = MSFiles::makeImageThumb(DOC_ROOT . $elem['path']['original'], $conf);
            $elem['path']['min'] = $result;
        }
    }

    $item['gallery'] = serialize($buf);

    $sql = 'UPDATE ' . PRFX . "catalog_items SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}
```

Правильное решение, хотя скорее обертка.

``` php
use pers1307\convert;

$convertImage = new ConvertImage();
$convertImage->setDocRoot(DOC_ROOT);
$convertImage->setConfig([115, 115, true]);
$convertImage->setCollectionKey('gallery');

$function = function($pathOriginal, $config) { MSFiles::makeImageThumb($pathOriginal, $config); };

$query = new MSTable('{catalog_items}');
$query->setFields(['*']);
$items = $query->getItems();

foreach ($items as $key => &$item) {

    $item = $convertImage->newImageInCollection($item, 'min', $function)

    $sql = 'UPDATE ' . PRFX . "catalog_items SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}
```

Пример переконвертирования картинки в галлерее картинок.
Быстрое решение

``` php
$query = new MSTable('{works}');
$query->setFields(['*']);
$items = $query->getItems();

$galleries = [];

foreach ($items as $key =>$item) {
    $arrayGallery = unserialize($item['gallery']);

    foreach ($arrayGallery as $key2 => $pic) {
        $galleries[$key][$key2] = $pic['path']['original'];
    }
}

$conf3 = array(800, 480,
    'watermark' => array(
        'src' => DOC_ROOT . '/DESIGN/SITE/images/watermark400x400.png',
        'offset_x' => 150,
        'offset_y' => 0
    )
);

foreach ($items as $key => $item) {

    if (isset($galleries[$key])) {

        $tempGal = unserialize($item['gallery']);

        foreach ($tempGal as $key2 => $temp) {

            // Переконфигурировать картинку
            //$galleries[$key][$key2];
            $result = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key][$key2], $conf3);

            $tempGal[$key2]['path']['win'] = $result;
        }

        $items[$key]['gallery'] = serialize($tempGal);
    }
}

foreach ($items as $key => $item) {
    $sql = 'UPDATE ' . PRFX . "works SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}
```

## Автор

- [Pereskokov Yurii (pers1307)](https://github.com/pers1307)

## Лицензия

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
#Changelog
All Notable changes to `pers1307/convert` will be documented in this file

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Remove
- Nothing

### Security
- Nothing

src/convert.php
<?php

require_once 'console.php';

/**
 * Добавление новой картинки в список
 */
$query = new MSTable('{catalog_items}');
$query->setFields(['*']);
$items = $query->getItems();

$conf = array(115, 115, true);

foreach ($items as $key => &$item) {
    $buf = unserialize($item['gallery']);

    foreach ($buf as $key2 => &$elem) {

        if (file_exists(DOC_ROOT . $elem['path']['original'])) {
            $result = MSFiles::makeImageThumb(DOC_ROOT . $elem['path']['original'], $conf);
            //$result = 'test /' . $elem['path']['original'];
            $elem['path']['min'] = $result;
        }
    }

    $item['gallery'] = serialize($buf);

    $sql = 'UPDATE ' . PRFX . "catalog_items SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}

src/convert2.php
<?php

require_once 'console.php';

// Перенос для каталога

$query = new MSTable('{catalog_items}');
$query->setFields(['*']);
$items = $query->getItems();

$imagies_out = [];
$galleries = [];

foreach ($items as $key =>$item) {
    $r = unserialize($item['image_out']);
    $t = unserialize($item['gallery']);

    if (isset($r[0]['path']['original'])) {
        $imagies_out[$key] = $r[0]['path']['original'];
    }

    if (isset($t[0]['path']['original'])) {
        $galleries[$key] = $t[0]['path']['original'];
    }
}

//$conf = array(160, 110, true,
//    'watermark' => array(
//        'src' => DOC_ROOT . '/DESIGN/SITE/images/watermark100x100.png',
//        'offset_x' => 30,
//        'offset_y' => 0
//    ));
//
//foreach ($items as $key => $item) {
//
//    if (isset($imagies_out[$key])) {
//        $result = MSFiles::makeImageThumb(DOC_ROOT . $imagies_out[$key], $conf);
//
//        $r = unserialize($item['image_out']);
//
//        $r[0]['path']['min'] = $result;
//
//        $r = serialize($r);
//
//        $items[$key]['image_out'] = $r;
//    }
//}
//
//foreach ($items as $key => $item) {
//    $sql = 'UPDATE ' . PRFX . "catalog_items SET `image_out`='" . $item['image_out'] . "' WHERE `id`=" . $item['id'];
//    MSCore::db()->execute($sql);
//}

$conf = array(80, 80, true,
    'watermark' => array(
        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark50x50.png',
        'offset_x' => 15,
        'offset_y' => 10
    )
);

$conf2 = array(300, 200, true,
    'watermark' => array(
        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark200x200.png',
        'offset_x' => 50,
        'offset_y' => 0
    ));

$conf3 = array(720, 480, true,
    'watermark' => array(
        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark400x400.png',
        'offset_x' => 150,
        'offset_y' => 0
    ));

foreach ($items as $key => $item) {

    if (isset($galleries[$key])) {
        $result = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf);

        $result2 = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf2);

        $result3 = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf3);


        $r = unserialize($item['gallery']);

        $r[0]['path']['min'] = $result;
        $r[0]['path']['first'] = $result2;
        $r[0]['path']['win'] = $result3;

        $r = serialize($r);

        $items[$key]['gallery'] = $r;
    }
}

foreach ($items as $key => $item) {
    $sql = 'UPDATE ' . PRFX . "catalog_items SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}



//$count = 0;
//
//foreach ($items as $key =>$item) {
//
//    if (isset($imagies_out[$key]) && isset($galleries[$key])) {
//        ++$count;
//    }
//
//    $r = unserialize($item['image_out']);
//    $t = unserialize($item['gallery']);
//
//    if (isset($r[0]['path']['original'])) {
//        $imagies_out[$key] = $r[0]['path']['original'];
//    }
//
//    if (isset($t[0]['path']['original'])) {
//        $galleries[$key] = $t[0]['path']['original'];
//    }
//}

//print_r($count);
//print_r(count($items));

//$conf = array(80, 80, true,
//    'watermark' => array(
//        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark50x50.png',
//        'offset_x' => 15,
//        'offset_y' => 10
//    )
//);
//
//$conf2 = array(300, 200, true,
//    'watermark' => array(
//        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark200x200.png',
//        'offset_x' => 50,
//        'offset_y' => 0
//    ));
//
//$conf3 = array(720, 480, true,
//    'watermark' => array(
//        'src' => DOC_ROOT.'/DESIGN/SITE/images/watermark400x400.png',
//        'offset_x' => 150,
//        'offset_y' => 0
//    ));
//
//foreach ($items as $key => $item) {
//
//    if (isset($imagies_out[$key]) && !isset($galleries[$key])) {
//
//        $result = MSFiles::makeImageThumb(DOC_ROOT . $imagies_out[$key], $conf);
//
//        $result2 = MSFiles::makeImageThumb(DOC_ROOT . $imagies_out[$key], $conf2);
//
//        $result3 = MSFiles::makeImageThumb(DOC_ROOT . $imagies_out[$key], $conf3);
//
//        $r = unserialize($item['gallery']);
//
//        $r[0]['path']['min'] = $result;
//        $r[0]['path']['first'] = $result2;
//        $r[0]['path']['win'] = $result3;
//
//        $r = serialize($r);
//
//        $items[$key]['gallery'] = $r;
//    }
//
////    if (isset($galleries[$key])) {
////        $result = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf);
////
////        $result2 = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf2);
////
////        $result3 = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key], $conf3);
////
////
////        $r = unserialize($item['gallery']);
////
////        $r[0]['path']['min'] = $result;
////        $r[0]['path']['first'] = $result2;
////        $r[0]['path']['win'] = $result3;
////
////        $r = serialize($r);
////
////        $items[$key]['gallery'] = $r;
////    }
//}
//
//foreach ($items as $key => $item) {
//    $sql = 'UPDATE ' . PRFX . "catalog_items SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
//    MSCore::db()->execute($sql);
//}

src/ConvertImage.php
<?php
/**
 * ConvertImage.php
 * Класс помогает конвертировать картинки
 *
 * @author      Pereskokov Yurii
 * @version     0.1
 * @copyright   2016 Pereskokov Yurii
 * @license     Mediasite LLC
 * @link        http://www.mediasite.ru/
 */

namespace pers1307\convert;

use KoKoKo\assert\Assert;

class ConvertImage
{
    /**
     * @var string
     * Каталог сайта
     */
    protected $docRoot;

    /**
     * @var array
     * Конфигурация для новой картинки
     */
    protected $config;

    /**
     * @var string
     * Ключ коллекции в которой надо поменять коллекцию картинок
     */
    protected $collectionKey;

    /**
     * @param $docRoot string
     */
    public function setDocRoot($docRoot)
    {
        Assert::assert($docRoot, 'docRoot')->notEmpty()->string();

        $this->docRoot = $docRoot;
    }

    /**
     * @param $config array
     */
    public function setConfig($config)
    {
        Assert::assert($config, 'config')->isArray();

        $this->config = $config;
    }

    /**
     * @param $collectionKey string
     */
    public function setCollectionKey($collectionKey)
    {
        Assert::assert($collectionKey, 'collectionKey')->notEmpty()->string();

        $this->collectionKey = $collectionKey;
    }

    /**
     * Добавить новую картинку в коллекцию
     *
     * @param $item array
     * @param $newImageKey string
     * @param $functionConvert
     *
     * @return array
     */
    public function newImageInCollection($item, $newImageKey, $functionConvert)
    {
        Assert::assert($newImageKey, 'newImageKey')->notEmpty()->string();

        $buf = unserialize($item[$this->collectionKey]);

        foreach ($buf as $key2 => &$elem) {

            if (file_exists($this->docRoot . $elem['path']['original'])) {

                $result = $functionConvert($this->docRoot . $elem['path']['original'], $this->config);
                $elem['path'][$newImageKey] = $result;
            }
        }

        $item[$this->collectionKey] = serialize($buf);

        return $item;
    }
}

src/convert_works.php
<?php

require_once 'console.php';

$query = new MSTable('{works}');
$query->setFields(['*']);
$items = $query->getItems();

$galleries = [];

foreach ($items as $key =>$item) {
    $arrayGallery = unserialize($item['gallery']);

    foreach ($arrayGallery as $key2 => $pic) {
        $galleries[$key][$key2] = $pic['path']['original'];
    }
}

$conf3 = array(800, 480,
    'watermark' => array(
        'src' => DOC_ROOT . '/DESIGN/SITE/images/watermark400x400.png',
        'offset_x' => 150,
        'offset_y' => 0
    )
);

foreach ($items as $key => $item) {

    if (isset($galleries[$key])) {

        $tempGal = unserialize($item['gallery']);

        foreach ($tempGal as $key2 => $temp) {

            // Переконфигурировать картинку
            //$galleries[$key][$key2];
            $result = MSFiles::makeImageThumb(DOC_ROOT . $galleries[$key][$key2], $conf3);

            $tempGal[$key2]['path']['win'] = $result;
        }

        $items[$key]['gallery'] = serialize($tempGal);
    }
}

foreach ($items as $key => $item) {
    $sql = 'UPDATE ' . PRFX . "works SET `gallery`='" . $item['gallery'] . "' WHERE `id`=" . $item['id'];
    MSCore::db()->execute($sql);
}